### PR TITLE
`merkledb` -- dynamic root

### PR DESCRIPTION
**Pull request by danlaine**
## Why this should be merged

The `root` field of a `trieView` or the database should be the actual `root`.
## How this works

* `root` is now a `maybe.Maybe[*node]` which is `Nothing` iff the trie is empty.
* A trie reports an `ids.Empty` root ID iff the trie is empty. 
* `changeSummary` now explicitly tracks the `rootChange`.
* In `trieview` the `root` is now the actual root of the trie.
* `getPathTo` may now return an empty path.
*  Prohibit empty proofs.

## How this was tested

Update existing and add new UT.

### CODE REVIEW

File: x/merkledb/trieview.go [hunk 1]
```diff
@@ -145,12 +145,9 @@ func newTrieView(
...
 ) (*trieView, error) {
-	root, err := parentTrie.getEditableNode(db.rootPath, false /* hasValue */)
-	if err != nil {
-		if err == database.ErrNotFound {
-			return nil, ErrNoValidRoot
-		}
-		return nil, err
+	root := parentTrie.getRoot()
+	if root.HasValue() {
+		root = maybe.Some(root.Value().clone()) // TODO better way of doing this?
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:150]: bind(root, clone)

File: x/merkledb/trieview.go [hunk 2]
```diff
@@ -662,27 +718,28 @@ func (t *trieView) remove(key Path) error {
 	return nil
 }
 
-// Merges together nodes in the inclusive descendants of [node] that
+// Merges together nodes in the inclusive descendnnts of [n] that
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:721]: spelling

File: x/merkledb/trieview.go [hunk 3]
```diff
@@ -692,22 +749,26 @@ func (t *trieView) compressNodePath(parent, node *node) error {
...
 	}
 
 	// [node] is the first node with multiple children.
 	// combine it with the [node] passed in.
-	parent.addChild(node)
+	if parent == nil {
+		t.root = maybe.Some[*node](n)
+		return nil
+	}
+	parent.addChild(n)
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:771]: isn't parent still nil here?

File: x/merkledb/trieview.go [hunk 4]
```diff
@@ -342,6 +358,10 @@ func (t *trieView) getProof(ctx context.Context, key []byte) (*Proof, error) {
 		return nil, err
 	}
 
+	if len(proofPath) == 0 {
+		return proof, nil
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:362]: should return the root since this is an exclusion proof

File: scripts/mocks.mockgen.txt [hunk 5]
```diff
@@ -49,5 +49,4 @@ github.com/ava-labs/avalanchego/vms/registry=VMGetter=vms/registry/mock_vm_gette
 github.com/ava-labs/avalanchego/vms/registry=VMRegisterer=vms/registry/mock_vm_registerer.go
 github.com/ava-labs/avalanchego/vms/registry=VMRegistry=vms/registry/mock_vm_registry.go
 github.com/ava-labs/avalanchego/vms=Factory,Manager=vms/mock_manager.go
-github.com/ava-labs/avalanchego/x/merkledb=MerkleDB=x/merkledb/mock_db.go
```
Comment (danlaine) [inline scripts/mocks.mockgen.txt:52]: Auto-generation fails because of generics 😭

File: x/merkledb/trieview.go (hunk 3 above)
Comment (danlaine) [reply x/merkledb/trieview.go:770]: idk if this comment is outdated now but I don't think `parent` can be `nil` here since we return in the `if` above

File: x/merkledb/db.go [hunk 6]
```diff
@@ -1242,12 +1269,43 @@ func (db *merkleDB) getNode(key Key, hasValue bool) (*node, error) {
...
-	return db.intermediateNodeDB.Get(key)
+
+	// Don't need to lock [view] because nobody else has a reference to it.
+	view, err := newTrieView(db, db, ViewChanges{BatchOps: ops})
+	if err != nil {
+		return err
+	}
+
+	return view.commitToDB(context.Background())
 }
```
Comment (StephenButtolph) [inline x/merkledb/db.go:1309]: Shouldn't we just be wiping the intermediate nodes and then wiping the value nodes? Or are we trying to allow someone to call this and maintain the change in the history?

File: x/merkledb/db.go [hunk 7]
```diff
@@ -1242,12 +1269,43 @@ func (db *merkleDB) getNode(key Key, hasValue bool) (*node, error) {
...
+	default:
+		return db.intermediateNodeDB.Get(key)
+	}
+}
+
+func (db *merkleDB) getRoot() maybe.Maybe[*node] {
+	return db.root
+}
+
+func (db *merkleDB) Clear() error {
```
Comment (StephenButtolph) [inline x/merkledb/db.go:1283]: Should we move the addition of `Clear` into a smaller PR? I don't think this really directly depends on the root change.

File: x/merkledb/proof.go [hunk 8]
```diff
@@ -840,9 +837,9 @@ func addPathInfo(
 			if existingChild, ok := n.children[index]; ok {
 				compressedKey = existingChild.compressedKey
 			}
-			childPath := key.Extend(ToToken(index, t.tokenSize), compressedKey)
-			if (shouldInsertLeftChildren && childPath.Less(insertChildrenLessThan.Value())) ||
-				(shouldInsertRightChildren && childPath.Greater(insertChildrenGreaterThan.Value())) {
+			childKey := key.Extend(ToToken(index, t.tokenSize), compressedKey)
+			if (shouldInsertLeftChildren && childKey.Less(insertChildrenLessThan.Value())) ||
+				(shouldInsertRightChildren && childKey.Greater(insertChildrenGreaterThan.Value())) {
```
Comment (StephenButtolph) [inline x/merkledb/proof.go:842]: If we pulled this rename into it's own PR I think we could merge it rq. Doesn't seem related to the root change.

File: x/merkledb/trie_test.go [hunk 9]
```diff
@@ -43,7 +43,7 @@ func getNodeValue(t ReadOnlyTrie, key string) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
-	if result.key != path || result == nil {
+	if result == nil || result.key != path {
```
Comment (StephenButtolph) [inline x/merkledb/trie_test.go:46]: Similar comment - I think we could pull this into a small fix PR

File: x/merkledb/trieview.go [hunk 10]
```diff
@@ -147,13 +147,9 @@ func newTrieView(
...
-	root, err := parentTrie.getEditableNode(Key{}, false /* hasValue */)
-	if err != nil {
-		if err == database.ErrNotFound {
-			return nil, ErrNoValidRoot
-		}
-		return nil, err
-	}
+	root := maybe.Bind(parentTrie.getRoot(), func(n *node) *node {
+		return n.clone()
+	})
```
Comment (StephenButtolph) [inline x/merkledb/trieview.go:152]:
```suggestion
	root := maybe.Bind(parentTrie.getRoot(), (*node).clone)
```

File: x/merkledb/trieview.go [hunk 11]
```diff
@@ -647,43 +672,78 @@ func (t *trieView) remove(key Key) error {
...
 
+	// Note [parent] != nil since [nodeToDelete.key] != [t.root.key].
+	// i.e. There's the root and at least one more node.
+
 	// if the removed node has no children, the node can be removed from the trie
-	if err := t.recordNodeDeleted(nodeToDelete); err != nil {
-		return err
-	}
-	if parent != nil {
+	// <<<<<<< HEAD
```
Comment (StephenButtolph) [inline x/merkledb/trieview.go:712]: Seems like a weird merge

File: x/sync/client.go [hunk 12]
```diff
@@ -268,8 +265,8 @@ func (c *client) GetRangeProof(
 			ctx,
 			&rangeProof,
 			int(req.KeyLimit),
-			startKey,
-			endKey,
+			maybeBytesToMaybe(req.StartKey),
+			maybeBytesToMaybe(req.EndKey),
```
Comment (StephenButtolph) [inline x/sync/client.go:269]: I think we could pull this into a separate small cleanup PR

File: x/sync/client_test.go [hunk 13]
```diff
@@ -339,11 +339,11 @@ func TestGetRangeProof(t *testing.T) {
 				BytesLimit: defaultRequestByteSizeLimit,
 			},
 			modifyResponse: func(response *merkledb.RangeProof) {
-				response.KeyValues = nil
 				response.StartProof = nil
 				response.EndProof = nil
+				response.KeyValues = nil
```
Comment (StephenButtolph) [inline x/sync/client_test.go:344]: nit could probably be in a separate cleanup PR (if at all)

**Review by StephenButtolph — COMMENTED** [review]
few suggestions for some non-controversial changes to pull into other PRs to make this a bit smaller. Still looking for if the number of special cases is actually less here than #2106

File: x/merkledb/proof.go (hunk 8 above)
Comment (danlaine) [reply x/merkledb/proof.go:842]: https://github.com/ava-labs/avalanchego/pull/2267

File: x/merkledb/trie_test.go (hunk 9 above)
Comment (danlaine) [reply x/merkledb/trie_test.go:46]: https://github.com/ava-labs/avalanchego/pull/2268

File: x/sync/client.go (hunk 12 above)
Comment (danlaine) [reply x/sync/client.go:269]: https://github.com/ava-labs/avalanchego/pull/2269

File: x/merkledb/db.go (hunk 7 above)
Comment (danlaine) [reply x/merkledb/db.go:1285]: https://github.com/ava-labs/avalanchego/pull/2277/files

**Comment by danlaine** [comment]
Hmm found a failing fuzz test. Looking into it. https://github.com/ava-labs/avalanchego/actions/runs/7064425097/job/19234865527#step:6:165

File: x/merkledb/db.go [hunk 14]
```diff
@@ -51,11 +51,11 @@ var (
 	intermediateNodePrefix = []byte{2}
 
 	cleanShutdownKey        = []byte(string(metadataPrefix) + "cleanShutdown")
+	rootDBKey               = []byte(string(metadataPrefix) + "root")
```
Comment (dboehm-avalabs) [inline x/merkledb/db.go:54]: realized belatedly that you don't actually need this since the root should always be the smallest key present in your nodes db.  You should be able to just grab the first key on the iterator, right?

File: x/merkledb/db.go [hunk 15]
```diff
@@ -1263,12 +1274,15 @@ func (db *merkleDB) getNode(key Key, hasValue bool) (*node, error) {
 	switch {
 	case db.closed:
 		return nil, database.ErrClosed
-	case key == Key{}:
```
Comment (dboehm-avalabs) [inline x/merkledb/db.go:1266]: you still need this for the root since you are keeping a copy of it in memory separate from the dbs

File: x/merkledb/trieview.go [hunk 16]
```diff
@@ -148,20 +147,12 @@ func newTrieView(
...
-		return nil, err
-	}
-
 	newView := &trieView{
-		sentinelNode: sentinelNode,
-		db:           db,
-		parentTrie:   parentTrie,
-		changes:      newChangeSummary(len(changes.BatchOps) + len(changes.MapOps)),
-		tokenSize:    db.tokenSize,
+		root:       maybe.Bind(parentTrie.getRoot(), (*node).clone),
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:151]: this no longer calls db.lock.RLock() before grabbing the root, not sure if it could be an issue since it still rlocks the commit lock, but requires more review.

File: x/merkledb/trieview.go [hunk 17]
```diff
@@ -231,6 +222,8 @@ func (t *trieView) calculateNodeIDs(ctx context.Context) error {
 		}
 		defer t.nodesAlreadyCalculated.Set(true)
 
+		oldRoot := maybe.Bind(t.root, (*node).clone)
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:225]: I don't think this needs to be cloned since before values don't get edited

File: x/merkledb/history.go [hunk 18]
```diff
@@ -249,6 +254,13 @@ func (th *trieHistory) getChangesToGetToRoot(rootID ids.ID, start maybe.Maybe[[]
 	for i := mostRecentChangeIndex; i > lastRootChangeIndex; i-- {
 		changes, _ := th.history.Index(i)
 
+		if i == mostRecentChangeIndex {
```
Comment (dboehm-avalabs) [inline x/merkledb/history.go:257]: should these just be outside the loop?

File: x/merkledb/db.go (hunk 14 above)
Comment (danlaine) [reply x/merkledb/db.go:54]: We could do that, but I feel like this is a bit cleaner than having to check both the value and intermediate node databases

File: x/merkledb/trieview.go (hunk 17 above)
Comment (danlaine) [reply x/merkledb/trieview.go:225]: I think we need to clone here because `t.root` can change during the `for key, change := range t.changes.values` loop. We can't just do `oldRoot := t.root` because we'll end up with `oldRoot` being set to the new root.

File: x/merkledb/history.go (hunk 18 above)
Comment (danlaine) [reply x/merkledb/history.go:262]: They could be, but then we'd end up having to get it from `th.history` again anyway, so I don't think that'd really buy us anything

File: x/merkledb/trieview.go [hunk 19]
```diff
@@ -655,43 +637,77 @@ func (t *trieView) remove(key Key) error {
...
+	if nodeToDelete.key == t.root.Value().key {
+		// We're deleting the root.
+		switch len(nodeToDelete.children) {
+		case 0:
+			// The trie is empty now.
+			t.root = maybe.Nothing[*node]()
+			return t.recordNodeDeleted(nodeToDelete)
+		case 1:
+			// The root has one child, so make that child the new root.
+			var (
```
[the remark this answers is not in the document]
Comment (danlaine) [reply x/merkledb/trieview.go]: good point, fixed

File: x/merkledb/db.go (hunk 15 above)
Comment (danlaine) [reply x/merkledb/db.go:1268]: updated

File: x/merkledb/trieview.go [hunk 20]
```diff
@@ -655,43 +636,50 @@ func (t *trieView) remove(key Key) error {
...
-	if err := t.recordNodeDeleted(nodeToDelete); err != nil {
-		return err
-	}
-	if parent != nil {
+	if len(nodeToDelete.children) == 0 {
+		if err := t.recordNodeDeleted(nodeToDelete); err != nil {
+			return err
+		}
+
+		if nodeToDelete.key == t.root.Value().key {
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:646]: Might want to move this into record node deleted since theoretically nodes get deleted from other functions as well.  We never actually delete the root from these other functions currently, but might be safer for future changes.

File: x/merkledb/trieview.go [hunk 21]
```diff
@@ -785,14 +786,54 @@ func (t *trieView) insert(
...
 	var closestNode *node
 	if err := t.visitPathToKey(key, func(n *node) error {
 		closestNode = n
+		// Need to recalculate ID for all nodes on path to [key].
 		return t.recordNodeChange(n)
 	}); err != nil {
 		return nil, err
 	}
 
+	if closestNode == nil {
```
[the remark this answers is not in the document]
Comment (dboehm-avalabs) [reply x/merkledb/trieview.go:807]: Except for the setting of t.root.

File: x/merkledb/trieview.go (hunk 20 above)
Comment (danlaine) [reply x/merkledb/trieview.go:646]: Nodes are only deleted in this method or its callees, so I'd be alright keeping this here, but I'll move it if you feel strongly

File: x/merkledb/db.go (hunk 14 above)
Comment (dboehm-avalabs) [reply x/merkledb/db.go:54]: I'm pretty sure what you have now is broken then, since you only record the key and not which db it is in.

File: x/merkledb/trieview.go (hunk 20 above)
Comment (dboehm-avalabs) [reply x/merkledb/trieview.go:646]: nah, not strongly.

**Review by dboehm-avalabs — APPROVED** [review]
